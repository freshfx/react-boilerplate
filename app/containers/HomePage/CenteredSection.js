import styled from '@emotion/styled'

import Section from './Section'

const CenteredSection = styled(Section)`
  text-align: center;
`

export default CenteredSection
