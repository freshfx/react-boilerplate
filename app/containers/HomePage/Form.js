import styled from '@emotion/styled'

const Form = styled.form`
  margin-bottom: 1em;
`

export default Form
