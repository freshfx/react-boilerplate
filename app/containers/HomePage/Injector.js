import {compose} from 'redux'
import {
  injectReducer,
  injectSaga
} from 'redux-injectors'

import reducer, {saga} from 'modules/repository/results'

/**
 * The Injectors have to use the HOCs (injectReducer and injectSaga)
 * instead of the hooks (useInjectReducer and useInjectSaga) until this issue is fixed:
 * https://github.com/react-boilerplate/redux-injectors/issues/19
 */
const HomePageInjector = () => null

export default compose(
  injectReducer(reducer),
  injectSaga(saga)
)(HomePageInjector)
