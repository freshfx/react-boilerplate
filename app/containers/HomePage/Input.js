import styled from '@emotion/styled'

const Input = styled.input`
  outline: none;
  border-bottom: 1px dotted #999;
`

export default Input
