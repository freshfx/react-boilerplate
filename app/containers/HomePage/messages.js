/*
 * HomePage Messages
 *
 * This contains all the text for the HomePage component.
 */
import {defineMessages} from 'react-intl'

export default defineMessages({
  startProjectHeader: {
    defaultMessage: 'Start your next react project in seconds',
    id: 'boilerplate.containers.HomePage.start_project.header'
  },
  startProjectMessage: {
    defaultMessage:
      'A highly scalable, offline-first foundation with the best DX and a focus on performance and best practices',
    id: 'boilerplate.containers.HomePage.start_project.message'
  },
  trymeAtPrefix: {
    defaultMessage: '@',
    id: 'boilerplate.containers.HomePage.tryme.atPrefix'
  },
  trymeHeader: {
    defaultMessage: 'Try me!',
    id: 'boilerplate.containers.HomePage.tryme.header'
  },
  trymeMessage: {
    defaultMessage: 'Show Github repositories by',
    id: 'boilerplate.containers.HomePage.tryme.message'
  }
})
