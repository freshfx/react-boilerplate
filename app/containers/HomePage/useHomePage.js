import {useInjectReducer} from 'redux-injectors'
import {
  useDispatch,
  useSelector
} from 'react-redux'

import reducer, {
  actions,
  selectors
} from 'modules/pages/home'

const useHomeEvents = () => {
  const dispatch = useDispatch()

  return {
    onChangeUsername: username =>
      dispatch(actions.changeUsername({username}))
  }
}

const useHomeState = () => ({
  username: useSelector(selectors.selectUsername)
})

const useHomePage = () => {
  useInjectReducer(reducer)

  const state = useHomeState()
  const events = useHomeEvents()

  return [
    state,
    events
  ]
}

export default useHomePage
