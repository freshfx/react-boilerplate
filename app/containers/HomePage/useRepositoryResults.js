import {useDispatch, useSelector} from 'react-redux'

import {actions, selectors} from 'modules/repository/results'

const useRepositoriesEvents = () => {
  const dispatch = useDispatch()

  return {
    onLoad: () => dispatch(actions.loadRepositories())
  }
}

const useRepositoriesState = () => ({
  error: useSelector(selectors.selectError),
  isLoading: useSelector(selectors.selectIsLoading),
  repositories: useSelector(selectors.selectRepositories)
})

const useRepositoryResults = () => {
  const state = useRepositoriesState()
  const events = useRepositoriesEvents()

  return [state, events]
}

export default useRepositoryResults
