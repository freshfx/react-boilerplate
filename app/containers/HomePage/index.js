/*
 * HomePage
 *
 * This is the first thing users see of our App, at the '/' route
 */

import React from 'react'
import PropTypes from 'prop-types'
import {Helmet} from 'react-helmet-async'
import {FormattedMessage} from 'react-intl'
import {connect} from 'react-redux'
import {compose} from 'redux'
import {createStructuredSelector} from 'reselect'
import noop from 'lodash/noop'
import {
  injectReducer,
  injectSaga
} from 'redux-injectors'

import H2 from 'components/H2'
import ReposList from 'components/ReposList'
import repositoriesReducer, {
  actions as repositoriesActions,
  saga as repositoriesSaga,
  selectors as repositoriesSelectors
} from 'modules/repository/results'
import reducer, {
  actions,
  selectors
} from 'modules/pages/home'

import AtPrefix from './AtPrefix'
import CenteredSection from './CenteredSection'
import Form from './Form'
import Input from './Input'
import Section from './Section'
import messages from './messages'

class HomePage extends React.PureComponent {
  /*
   * When initial state username is not null, submit the form to load repos
   */
  componentDidMount() {
    if (this.props.username && this.props.username.trim().length) {
      this.props.onSubmitForm()
    }
  }

  render() {
    return (
      <article>
        <Helmet>
          <title>Home Page</title>
          <meta
            name="description"
            content="A React.js Boilerplate application homepage"
          />
        </Helmet>
        <div>
          <CenteredSection>
            <H2>
              <FormattedMessage {...messages.startProjectHeader} />
            </H2>
            <p>
              <FormattedMessage {...messages.startProjectMessage} />
            </p>
          </CenteredSection>
          <Section>
            <H2>
              <FormattedMessage {...messages.trymeHeader} />
            </H2>
            <Form onSubmit={this.props.onSubmitForm}>
              <label htmlFor="username">
                <FormattedMessage {...messages.trymeMessage} />
                <AtPrefix>
                  <FormattedMessage {...messages.trymeAtPrefix} />
                </AtPrefix>
                <Input
                  id="username"
                  type="text"
                  placeholder="mxstbr"
                  value={this.props.username}
                  onChange={this.props.onChangeUsername}
                />
              </label>
            </Form>
            <ReposList
              error={this.props.error}
              loading={this.props.isLoading}
              repos={this.props.repositories}
            />
          </Section>
        </div>
      </article>
    )
  }
}

HomePage.propTypes = {
  error: PropTypes.oneOfType([
    PropTypes.object,
    PropTypes.bool
  ]),
  isLoading: PropTypes.bool,
  onChangeUsername: PropTypes.func,
  onSubmitForm: PropTypes.func,
  repositories: PropTypes.oneOfType([
    PropTypes.array,
    PropTypes.bool
  ]),
  username: PropTypes.string
}

HomePage.defaultProps = {
  error: false,
  isLoading: false,
  onChangeUsername: noop,
  onSubmitForm: noop,
  repositories: false,
  username: ''
}

const mapStateToProps = createStructuredSelector({
  error: repositoriesSelectors.selectError,
  isLoading: repositoriesSelectors.selectIsLoading,
  repositories: repositoriesSelectors.selectRepositories,
  username: selectors.selectUsername
})

const mapDispatchToProps = dispatch => ({
  onChangeUsername: event =>
    dispatch(actions.changeUsername({username: event.target.value})),
  onSubmitForm: event => {
    if (typeof event !== 'undefined' && event.preventDefault) {
      event.preventDefault()
    }
    dispatch(repositoriesActions.loadRepositories())
  }
})

const withConnect = connect(mapStateToProps, mapDispatchToProps)
const withReducer = injectReducer(reducer)
const withRepositoriesReducer = injectReducer(repositoriesReducer)
const withRepositoriesSaga = injectSaga(repositoriesSaga)

export {
  HomePage,
  mapDispatchToProps
}

export default compose(
  withConnect,
  withReducer,
  withRepositoriesReducer,
  withRepositoriesSaga
)(HomePage)
