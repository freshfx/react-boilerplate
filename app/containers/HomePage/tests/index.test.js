/**
 * Test the HomePage
 */

import React from 'react'
import {
  fireEvent,
  render
} from '@testing-library/react'
import {IntlProvider} from 'react-intl'
import {HelmetProvider} from 'react-helmet-async'
import {Provider} from 'react-redux'

import configureStore from 'configure-store'
import {
  actions
} from 'modules/pages/home'
import {
  actions as repositoriesActions
} from 'modules/repository/results'
import history from 'utils/history'

import HomePage from '../index'
import messages from '../messages'

jest.mock('components/ReposList', () => props => {
  if (props.loading) {
    return 'ReposList Loading'
  }
  if (props.error) {
    return 'ReposList Error'
  }
  if (props.repos.length) {
    return 'ReposList'
  }
  return 'ReposList Empty'
})

const store = configureStore({}, history)
const dispatch = jest.spyOn(store, 'dispatch')

const renderComponent = () =>
  render(
    <IntlProvider locale="en">
      <HelmetProvider>
        <Provider store={store}>
          <HomePage />
        </Provider>
      </HelmetProvider>
    </IntlProvider>
  )

describe('HomePage', () => {
  beforeEach(() => {
    dispatch.mockClear()
  })

  it('should match the snapshot', () => {
    const {container} = renderComponent()
    expect(container.firstChild).toMatchSnapshot()
  })

  it('should render the repo list with the correct props', () => {
    const {getByText} = renderComponent()
    expect(getByText('ReposList Empty')).toBeInTheDocument()

    store.dispatch(repositoriesActions.loadRepositories())
    expect(getByText('ReposList Loading')).toBeInTheDocument()

    const repositories = [1, 2, 3]
    store.dispatch(repositoriesActions.repositoriesLoaded({repositories}))
    expect(getByText('ReposList')).toBeInTheDocument()

    store.dispatch(repositoriesActions.repositoriesLoadingError({error: {}}))
    expect(getByText('ReposList Error')).toBeInTheDocument()
  })

  it('should fetch the repos on mount if a username exists', () => {
    store.dispatch(actions.changeUsername({username: 'mxstbr'}))
    renderComponent()
    expect(store.dispatch).toHaveBeenCalledWith(repositoriesActions.loadRepositories())
    store.dispatch(actions.changeUsername({username: ''}))
  })

  it('should not fetch the repos on mount', () => {
    renderComponent()
    expect(store.dispatch).not.toHaveBeenCalled()
  })

  it('should call the correct event when a user changes the input', () => {
    const username = 'mxstbr'
    const {getByLabelText} = renderComponent()
    const domNode = getByLabelText(RegExp(messages.trymeMessage.defaultMessage, 'u'))
    const event = {target: {value: username}}
    fireEvent.change(domNode, event)

    expect(store.dispatch).toHaveBeenCalledTimes(1)
    expect(store.dispatch).toHaveBeenCalledWith(actions.changeUsername({username}))

    store.dispatch(actions.changeUsername({username: ''}))
  })

  it('should call the correct event when a user submits the form', () => {
    const {getByPlaceholderText} = renderComponent()
    fireEvent.submit(getByPlaceholderText('mxstbr'))

    expect(store.dispatch).toHaveBeenCalledTimes(1)
    expect(store.dispatch).toHaveBeenCalledWith(repositoriesActions.loadRepositories())
  })
})
