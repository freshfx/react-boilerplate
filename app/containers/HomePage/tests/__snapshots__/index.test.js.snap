// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomePage should render the repos list 1`] = `
.emotion-2 {
  margin: 3em auto;
  text-align: center;
}

.emotion-2:first-of-type {
  margin-top: 0;
}

.emotion-0 {
  font-size: 1.5em;
}

.emotion-12 {
  margin: 3em auto;
}

.emotion-12:first-of-type {
  margin-top: 0;
}

.emotion-10 {
  margin-bottom: 1em;
}

.emotion-6 {
  color: black;
  margin-left: 0.4em;
}

.emotion-8 {
  outline: none;
  border-bottom: 1px dotted #999;
}

<article>
  <div>
    <section
      class="emotion-2 emotion-3"
    >
      <h2
        class="emotion-0 emotion-1"
      >
        Start your next react project in seconds
      </h2>
      <p>
        A highly scalable, offline-first foundation with the best DX and a focus on performance and best practices
      </p>
    </section>
    <section
      class="emotion-12 emotion-13"
    >
      <h2
        class="emotion-0 emotion-1"
      >
        Try me!
      </h2>
      <form
        class="emotion-10 emotion-11"
      >
        <label
          for="username"
        >
          Show Github repositories by
          <span
            class="emotion-6 emotion-7"
          >
            @
          </span>
          <input
            class="emotion-8 emotion-9"
            id="username"
            placeholder="mxstbr"
            type="text"
            value=""
          />
        </label>
      </form>
      <div>
        Repos List
      </div>
    </section>
  </div>
</article>
`;
